Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -315,311 +545,498 @@ void CGContextSetFillColorSpace(CGContextRef pContext, CGColorSpaceRef colorSpac
+void CGContextSetGrayFillColor(CGContextRef context, CGFloat gray, CGFloat alpha) {
+    CGContextSetRGBFillColor(context, gray, gray, gray, alpha);
 }
 
 /**
  @Status Interoperable
 */
-CGPoint CGContextGetTextPosition(CGContextRef ctx) {
-    CGPoint ret;
-    ctx->Backing()->CGContextGetTextPosition(&ret);
+void CGContextSetRGBFillColor(CGContextRef context, CGFloat r, CGFloat g, CGFloat b, CGFloat a) {
+    ComPtr<ID2D1SolidColorBrush> brush;
+    FAIL_FAST_IF_FAILED(context->RenderTarget()->CreateSolidColorBrush({ r, g, b, a }, &brush));
+    FAIL_FAST_IF_FAILED(brush.As(&context->CurrentGState().fillBrush));
+}
 
-    return ret;
+/**
+ @Status Caveat
+ @Notes Manually converts CMYK to RGB, and does not involve the colorspace.
+*/
+void CGContextSetCMYKFillColor(CGContextRef context, CGFloat cyan, CGFloat magenta, CGFloat yellow, CGFloat black, CGFloat alpha) {
+    CGContextSetRGBFillColor(context,
+                             (1.0f - cyan) * (1.0f - black),
+                             (1.0f - magenta) * (1.0f - black),
+                             (1.0f - yellow) * (1.0f - black),
+                             alpha);
 }
+#pragma endregion
 
+#pragma region Graphics States
 /**
  @Status Interoperable
 */
-void CGContextSaveGState(CGContextRef ctx) {
-    ctx->Backing()->CGContextSaveGState();
+void CGContextSaveGState(CGContextRef context) {
+    auto& oldState = context->CurrentGState();
+
+    // This uses the drawing state's copy constructor.
+    context->GStateStack().emplace(oldState);
+
+    auto factory = context->Factory();
+    ComPtr<ID2D1DrawingStateBlock> drawingState;
+    FAIL_FAST_IF_FAILED(factory->CreateDrawingStateBlock(&drawingState));
+
+    context->RenderTarget()->SaveDrawingState(drawingState.Get());
+    oldState.d2dState = drawingState;
 }
 
 /**
  @Status Interoperable
 */
-void CGContextRestoreGState(CGContextRef ctx) {
-    ctx->Backing()->CGContextRestoreGState();
+void CGContextRestoreGState(CGContextRef context) {
+    if (context->GStateStack().size() <= 1) {
+        TraceError(TAG, L"Invalid attempt to pop last graphics state.");
+        return;
+    }
+
+    context->GStateStack().pop();
+    auto& newState = context->CurrentGState();
+    context->RenderTarget()->RestoreDrawingState(newState.d2dState.Get());
+    newState.d2dState = nullptr;
 }
+#pragma endregion
 
+#pragma region Shape Drawing - Operations
 /**
  @Status Interoperable
 */
-void CGContextClearRect(CGContextRef ctx, CGRect rct) {
-    if (!ctx) {
-        return;
+void CGContextClearRect(CGContextRef context, CGRect rect) {
+    ComPtr<ID2D1RenderTarget> renderTarget = context->RenderTarget();
+    renderTarget->PushAxisAlignedClip(__CGRectToD2D(rect), D2D1_ANTIALIAS_MODE_PER_PRIMITIVE);
+    renderTarget->Clear(nullptr); // transparent black clear
+    renderTarget->PopAxisAlignedClip();
+}
+
+static void __CGContextDrawGeometry(CGContextRef context, ID2D1Geometry* geometry, CGPathDrawingMode drawMode) {
```

Comment:
Absolutely. That would be ideal. :D

The effect stack will ideally consume an "image" input, which could be a command list or a bitmap or what-have-you.

---
In reply to: [82921442](https://github.com/Microsoft/WinObjC/pull/1126#discussion_r82921442) [](ancestors = 82921442)